cylc/uiserver/resolvers.py

```diff
@@ -321,6 +324,76 @@ async def play(cls, workflows, args, workflows_mgr, log):
         await workflows_mgr.scan()
         return response
 
+    @classmethod
+    async def enqueue(cls, stream, queue):
+        async for line in stream:
+            await queue.put(line.decode())
+
+    @classmethod
+    async def cat_log(cls, workflow: Tokens, log, info, task=None, file=None):
+        """Calls `cat log`.
+
+        Used for log subscriptions.
+        """
+        full_workflow = workflow
+        if file and task:
+            full_workflow = Tokens(f"{workflow.workflow_id}//{task}")
+        cmd = ['cylc', 'cat-log']
+        if file:
+            cmd += ['-f', file]
+        else:
+            cmd += ['-m', 't']
+        cmd.append(full_workflow.id)
+        log.info(f'$ {" ".join(cmd)}')
+
+        # For info, below subprocess is safe (uses shell=false by default)
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            *cmd,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+        buffer: List[str] = []
+        queue: asyncio.Queue = asyncio.Queue()
+        # Farm out reading from stdout stream to a background task
+        # This is to get around problem where stream is not EOF until
+        # subprocess ends
+        enqueue_task = asyncio.create_task(cls.enqueue(proc.stdout, queue))
+        op_id = info.root_value
+        if op_id not in info.context['sub_statuses']:
+            # subscription started in graphiql, add to dict
+            info.context['sub_statuses'][op_id] = 'start'
+        while True:
+            if (
+                op_id in info.context['sub_statuses']
+                and info.context['sub_statuses'][op_id] == 'stop'
+            ):
+                with suppress(psutil.Error):
+                    cat_log_proc = psutil.Process(proc.pid)
+                    # cat-log process will terminate on tail termination
+                    for tail_proc in cat_log_proc.children():
+                        tail_proc.terminate()
+                    enqueue_task.cancel()
+                    with suppress(asyncio.CancelledError):
+                        await enqueue_task
+                break
+            if queue.empty():
+                if buffer:
+                    yield list(buffer)
+                    buffer.clear()
+                if proc.returncode not in [None, 0]:
+                    (_, stderr) = await proc.communicate()
+                    # pass any error onto ui
+                    yield [stderr.decode()]
+                    break
+                await asyncio.sleep(0.1)
```

Comment:
This 0.1 second sleep was a placeholder value, I think we will want to increase this for production.

I found this in `man tail`:

```
       -s, --sleep-interval=N
              with  -f,  sleep  for  approximately N seconds (default 1.0) between
              iterations; with inotify and --pid=P, check process P at least  once
              every N seconds
```

Which suggests that new log lines will only come through at one second intervals so we might as well set this sleep to one second.